### PR TITLE
Handling for other return codes

### PR DESCRIPTION
**Pull request by GavinDmello**
This has the extra error codes Aedes doesn't return

### CODE REVIEW

File: lib/handlers/connect.js [hunk 1]
```diff
@@ -63,10 +63,44 @@ function authenticate (arg, done) {
...
+            break
+
+          case 4 :
+          default :
+            write(client, {
+              cmd: 'connack',
+              returnCode: 4
+            }, client.close.bind(client, done))
+            break
+      }
```
Comment (mcollina) [inline lib/handlers/connect.js]: how about a check on err.returnCode if it's between 1 and 4, and use 4 instead? We can avoid the switch.

**Comment by mcollina** [comment]
Can you please add unit tests and docs?

**Comment by GavinDmello** [comment]
Cool. I'll make the changes :)

**Comment by GavinDmello** [comment]
Will there be test cases for each return code , or a generic one ? Also do I have to add documentation in the Readme ?

**Comment by mcollina** [comment]
I think you should have test cases for each return code, yes. You can set them up via a for loop.

Yes, documentation goes in the README.

Thanks!

File: lib/handlers/connect.js [hunk 2]
```diff
@@ -63,10 +63,26 @@ function authenticate (arg, done) {
...
-        cmd: 'connack',
-        returnCode: 4
-      }, client.close.bind(client, done))
+      if (err && !err.returnCode) {
+        write(client, {
+          cmd: 'connack',
+          returnCode: 4
+        }, client.close.bind(client, done))
+        return
+      } else {
```
Comment (mcollina) [inline lib/handlers/connect.js]: You can probably avoid the first if.

Comment (GavinDmello) [inline lib/handlers/connect.js]: That was for checking if somebody doesn't send the returnCode property. I can merge it with the if below though.

Comment (mcollina) [inline lib/handlers/connect.js]: that's what I meant :)

File: README.md [hunk 3]
```diff
@@ -163,6 +163,19 @@ instance.authenticate = function (client, username, password, callback) {
   callback(null, username === 'matteo')
 }
 ```
+Other return codes can passed as follows :-
+
+```js
+instance.authenticate = function (client, username, password, callback) {
+  callback({ returnCode: 1 }, null)
+}
+```
```
Comment (mcollina) [inline README.md]: Can you please attach the returnCode to an Error object?

File: README.md [hunk 4]
```diff
@@ -163,6 +163,19 @@ instance.authenticate = function (client, username, password, callback) {
...
+The return code values and their responses which can be passed are given below :-
+
+1 - Unacceptable protocol version
+2 - Identifier rejected
+3 - Server unavailable
+4 - Bad user name or password
```
Comment (mcollina) [inline README.md]: Can you please make this bullet points, and wrap the numbers with backtick?

File: README.md (hunk 3 above)
Comment (GavinDmello) [inline README.md]: Do you mean assigning it to an error variable ?

**Comment by mcollina** [comment]
That's perfect, thanks very much!

**Pull request merged**